sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java

```diff
@@ -164,146 +152,121 @@ public void forceFlush() {
-      this.maxQueueSize = maxQueueSize;
-      this.halfMaxQueueSize = maxQueueSize >> 1;
+      this.scheduleDelayNanos = TimeUnit.MILLISECONDS.toNanos(scheduleDelayMillis);
       this.maxExportBatchSize = maxExportBatchSize;
       this.exporterTimeoutMillis = exporterTimeoutMillis;
-      this.spansList = new ArrayList<>(maxQueueSize);
+      this.queue = queue;
+      this.batch = new ArrayList<>(this.maxExportBatchSize);
     }
 
     private void addSpan(ReadableSpan span) {
-      synchronized (monitor) {
-        if (spansList.size() == maxQueueSize) {
-          droppedSpans.add(1);
-          return;
-        }
-        // TODO: Record a gauge for referenced spans.
-        spansList.add(span);
-        // Notify the worker thread that at half of the queue is available. It will take
-        // time anyway for the thread to wake up.
-        if (spansList.size() >= halfMaxQueueSize) {
-          monitor.notifyAll();
-        }
+      if (!queue.offer(span)) {
+        droppedSpans.add(1);
       }
     }
 
     @Override
     public void run() {
-      while (!Thread.currentThread().isInterrupted()) {
-        // Copy all the batched spans in a separate list to release the monitor lock asap to
-        // avoid blocking the producer thread.
-        ArrayList<ReadableSpan> spansCopy;
-        synchronized (monitor) {
-          // If still maxExportBatchSize elements in the queue better to execute an extra
-          if (spansList.size() < maxExportBatchSize) {
-            do {
-              // In the case of a spurious wakeup we export only if we have at least one span in
-              // the batch. It is acceptable because batching is a best effort mechanism here.
-              try {
-                monitor.wait(scheduleDelayMillis);
-              } catch (InterruptedException ie) {
-                // Preserve the interruption status as per guidance and stop doing any work.
-                Thread.currentThread().interrupt();
-                return;
-              }
-            } while (spansList.isEmpty());
+      updateNextExportTime();
+
+      while (continueWork) {
+        try {
+          if (flushRequested != null) {
+            flush();
+          }
+
+          ReadableSpan lastElement = queue.poll(100, TimeUnit.MILLISECONDS);
+          if (lastElement != null) {
+            batch.add(lastElement.toSpanData());
+          }
+
+          if (batch.size() >= maxExportBatchSize || System.nanoTime() >= nextExportTime) {
+            exportCurrentBatch();
+            updateNextExportTime();
           }
-          spansCopy = new ArrayList<>(spansList);
-          spansList.clear();
+
+        } catch (InterruptedException e) {
+          Thread.currentThread().interrupt();
+          return;
+        }
+      }
+    }
+
+    private void flush() {
+      int spansToFlush = queue.size();
+      while (spansToFlush > 0) {
+        ReadableSpan span = queue.poll();
```

Comment:
Similar to above, it might be better if we can structure this to take advantage of drainTo. I guess you can consecutively drainTo and subtract the number drained from spansToFlush instead of individual poll without a huge change in structure.

Reply:
I will try to benchmark this class today

Reply:
See https://github.com/open-telemetry/opentelemetry-java/pull/1579#discussion_r475713805